engine/src/flutter/impeller/renderer/backend/gles/blit_command_gles.cc

```diff
@@ -328,10 +330,29 @@ bool BlitCopyTextureToBufferCommandGLES::Encode(
   }
 
   DeviceBufferGLES::Cast(*destination)
-      .UpdateBufferData([&gl, this](uint8_t* data, size_t length) {
-        gl.ReadPixels(source_region.GetX(), source_region.GetY(),
-                      source_region.GetWidth(), source_region.GetHeight(),
-                      GL_RGBA, GL_UNSIGNED_BYTE, data + destination_offset);
+      .UpdateBufferData([&gl, this, coord_system,
+                         rows = source->GetSize().height](uint8_t* data,
+                                                          size_t length) {
+        switch (coord_system) {
+          case TextureCoordinateSystem::kUploadFromHost:
+            gl.ReadPixels(source_region.GetX(), source_region.GetY(),
+                          source_region.GetWidth(), source_region.GetHeight(),
+                          GL_RGBA, GL_UNSIGNED_BYTE, data + destination_offset);
+            break;
+          case TextureCoordinateSystem::kRenderToTexture:
+            // The texture is upside down, and must be inverted when copying
+            // byte data out.
+            size_t offset = destination_offset;
```

Comment:
I have nothing to provide as evidence for this other than perhaps an irrational fear that multiple calls to `ReadPixels` will make some driver do the most insane thing possible. But can we read in one shot and then flip the image in place after instead?

Iterate over the rows picking two from each end and `std::swap_ranges` on each perhaps.

Reply:
This is a reasonable fear, but I also doubt my ability to correctly use std::swap_ranges so it might be 6 of 1 and half a dozen of another...

Reply:
```c++
static void FlipImage(uint8_t* buffer,
                      size_t width,
                      size_t height,
                      size_t stride) {
  if (buffer == nullptr || stride == 0) {
    return;
  }

  const auto byte_width = width * stride;

  for (size_t top = 0; top < height; top++) {
    size_t bottom = height - top - 1;
    if (top >= bottom) {
      break;
    }
    auto* top_row = buffer + byte_width * top;
    auto* bottom_row = buffer + byte_width * bottom;
    std::swap_ranges(top_row, top_row + byte_width, bottom_row);
  }
}
```

Reply:
Will update the PR to use this.